pinot-broker/src/main/java/org/apache/pinot/broker/routing/v2/instanceselector/BaseInstanceSelector.java

```diff
@@ -0,0 +1,177 @@
+
+  private final AtomicLong _requestId = new AtomicLong();
+  private final String _tableNameWithType;
+  private final BrokerMetrics _brokerMetrics;
+
+  private volatile Set<String> _enabledInstances;
+  private volatile Map<String, List<String>> _segmentToInstancesMap;
+  private volatile Map<String, List<String>> _instanceToSegmentsMap;
+  private volatile Map<String, List<String>> _segmentToEnabledInstancesMap;
+
+  BaseInstanceSelector(String tableNameWithType, BrokerMetrics brokerMetrics) {
+    _tableNameWithType = tableNameWithType;
+    _brokerMetrics = brokerMetrics;
+  }
+
+  @Override
+  public void init(Set<String> enabledInstances, ExternalView externalView, Set<String> onlineSegments) {
+    _enabledInstances = enabledInstances;
+    onExternalViewChange(externalView, onlineSegments);
+  }
+
+  @Override
+  public void onInstancesChange(Set<String> enabledInstances, List<String> changedInstances) {
+    _enabledInstances = enabledInstances;
+
+    // Update all segments served by the changed instances
+    Set<String> segmentsToUpdate = new HashSet<>();
+    Map<String, List<String>> instanceToSegmentsMap = _instanceToSegmentsMap;
+    for (String instance : changedInstances) {
+      List<String> segments = instanceToSegmentsMap.get(instance);
+      if (segments != null) {
+        segmentsToUpdate.addAll(segments);
+      }
+    }
+
+    // Directly return if no segment needs to be updated
+    if (segmentsToUpdate.isEmpty()) {
+      return;
+    }
+
+    // Generate a new map from segment to enabled instances
+    Map<String, List<String>> segmentToInstancesMap = _segmentToInstancesMap;
+    Map<String, List<String>> segmentToEnabledInstancesMap = _segmentToEnabledInstancesMap;
+    Map<String, List<String>> newSegmentToEnabledInstancesMap =
+        new HashMap<>(HashUtil.getHashMapCapacity(segmentToEnabledInstancesMap.size()));
+    for (Map.Entry<String, List<String>> entry : segmentToEnabledInstancesMap.entrySet()) {
+      String segment = entry.getKey();
+      if (segmentsToUpdate.contains(segment)) {
+        newSegmentToEnabledInstancesMap.put(segment,
+            calculateEnabledInstancesForSegment(segment, segmentToInstancesMap.get(segment), enabledInstances));
+      } else {
+        newSegmentToEnabledInstancesMap.put(segment, entry.getValue());
+      }
+    }
+    _segmentToEnabledInstancesMap = newSegmentToEnabledInstancesMap;
+  }
+
+  @Override
+  public void onExternalViewChange(ExternalView externalView, Set<String> onlineSegments) {
+    Map<String, Map<String, String>> segmentAssignment = externalView.getRecord().getMapFields();
+    Map<String, List<String>> segmentToInstancesMap =
+        new HashMap<>(HashUtil.getHashMapCapacity(segmentAssignment.size()));
+    Map<String, List<String>> instanceToSegmentsMap = new HashMap<>();
+
+    for (Map.Entry<String, Map<String, String>> entry : segmentAssignment.entrySet()) {
+      // Keep segments in both external view and ideal state
```

Comment:
I see some very similar code from `Offline/Realtime Segment Selector`. It's inevitable to put on both?

Reply:
Yes, as the realtime one needs to handle CONSUMING state

Reply:
We don't need to skip segments not in the ideal state here. Updated